Sources/ClientRuntime/Interceptor/InterceptorContext.swift

```diff
@@ -0,0 +1,172 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/// The base type of all context objects passed to `Interceptor` methods.
+public protocol InterceptorContext {
+
+    /// The type of the transport message that will be transmitted by the operation being invoked.
+    associatedtype RequestType
+
+    /// The type of the transport message that will be received by the operation being invoked.
+    associatedtype ResponseType
+
+    /// The type of the attributes that will be available to all interceptors.
+    associatedtype AttributesType: HasAttributes
+
+    /// - Returns: The input for the operation being invoked.
+    func getInput() -> Any
```

Comment:
Isn't the type of the input `RequestType`?  Not sure why this return value is untyped.

Reply:
`input` and `output` are the types of the modeled operation input and output, respectively. `RequestType` and `ResponseType` are the transport message types, like `SdkHttpRequest`, `HttpResponse`

Reply:
Should the modeled input type be an associated type of this protocol like the RequestType & ResponseType then?  That way you can return the actual modeled input type, and you don't need the `func getInput<T>() -> T?` method below.

Reply:
Done